File: backend/prisma/seed.js
import { PrismaClient } from '@prisma/client';
import bcrypt from 'bcryptjs';

const prisma = new PrismaClient();

// Niveaux de difficulté avec multiplicateurs
const DIFFICULTIES = {
    EASY: { name: 'EASY', multiplier: 1.0 },
    MEDIUM: { name: 'MEDIUM', multiplier: 1.5 },
    HARD: { name: 'HARD', multiplier: 2.0 },
};

// Calcul du niveau basé sur l'XP total
function calculateLevel(totalXp) {
    return Math.floor(totalXp / 1000) + 1;
}

/**
 * Crée des sessions d'entraînement réalistes avec des dates garanties dans la semaine
 */
async function createSessionsAndStats(userId, sessionCount, streakDays, userName) {
    const sessions = [];
    const now = new Date();
    let totalXp = 0;

    for (let i = 0; i < sessionCount; i++) {
        // Garantir que 70% des sessions sont dans les 7 derniers jours
        // avec au moins 1 session par jour de la semaine
        let daysAgo;

        if (i < 7) {
            // Les 7 premières sessions : une par jour de la semaine
            daysAgo = i;
        } else if (i < sessionCount * 0.7) {
            // Sessions supplémentaires dans la semaine
            daysAgo = Math.floor(Math.random() * 7);
        } else {
            // Le reste dans les 2 semaines précédentes
            daysAgo = Math.floor(Math.random() * 14) + 7;
        }

        // Heure réaliste (8h-20h)
        const hours = Math.floor(Math.random() * 12) + 8;
        const minutes = Math.floor(Math.random() * 60);

        const completedAt = new Date(now);
        completedAt.setDate(completedAt.getDate() - daysAgo);
        completedAt.setHours(hours, minutes, 0, 0);

        // Choix de la difficulté (pondéré)
        const diffRoll = Math.random();
        let difficulty;
        if (diffRoll < 0.35) difficulty = DIFFICULTIES.EASY;
        else if (diffRoll < 0.75) difficulty = DIFFICULTIES.MEDIUM;
        else difficulty = DIFFICULTIES.HARD;

        // Données réalistes
        const totalImages = Math.floor(Math.random() * 10) + 10; // 10-20 images
        const correctAnswers = Math.floor(totalImages * (0.6 + Math.random() * 0.35)); // 60-95%
        const precision = Math.round((correctAnswers / totalImages) * 100 * 10) / 10;
        const duration = Math.floor(Math.random() * 300) + 120; // 2-7 minutes

        // Calcul du score et XP
        const baseScore = Math.floor(precision * 10);
        const xpEarned = Math.round(baseScore * difficulty.multiplier);
        totalXp += xpEarned;

        const session = await prisma.trainingSession.create({
            data: {
                userId,
                difficulty: difficulty.name,
                precision,
                duration,
                totalImages,
                correctAnswers,
                baseScore,
                multiplier: difficulty.multiplier,
                xpEarned,
                completedAt,
            },
        });
        sessions.push(session);
    }
    console.log(`✅ Created ${sessions.length} sessions for ${userName} (${totalXp} XP)`);

    // Statistiques
    const totalSessions = sessions.length;
    const averageScore = Math.round(sessions.reduce((acc, s) => acc + s.precision, 0) / totalSessions);
    const averageTime = Math.round(sessions.reduce((acc, s) => acc + s.duration, 0) / totalSessions);
    const level = calculateLevel(totalXp);

    await prisma.userStats.upsert({
        where: { userId },
        update: { totalXp, level, totalSessions, averageScore, averageTime, currentStreak: streakDays, lastActivityAt: now },
        create: { userId, totalXp, level, totalSessions, averageScore, averageTime, currentStreak: streakDays, lastActivityAt: now },
    });
    console.log(`✅ Stats: Level ${level}, ${totalXp} XP, ${averageScore}% avg`);
}

async function main() {
    console.log('🌱 Seeding database...');

    // Admin - 30 sessions
    const adminPassword = await bcrypt.hash('admin123', 10);
    const admin = await prisma.user.upsert({
        where: { email: 'admin@eroz.com' },
        update: {},
        create: { email: 'admin@eroz.com', password: adminPassword, firstName: 'Admin', lastName: 'Éroz', role: 'ADMIN' },
    });
    console.log('✅ Admin:', admin.email);
    await createSessionsAndStats(admin.id, 30, 15, admin.firstName);

    // Student - 12 sessions
    const studentPassword = await bcrypt.hash('student123', 10);
    const student = await prisma.user.upsert({
        where: { email: 'thomas.martin@edu.fr' },
        update: {},
        create: { email: 'thomas.martin@edu.fr', password: studentPassword, firstName: 'Thomas', lastName: 'Martin', role: 'STUDENT' },
    });
    console.log('✅ Student:', student.email);
    await createSessionsAndStats(student.id, 12, 5, student.firstName);

    // Professor - 20 sessions
    const profPassword = await bcrypt.hash('prof123', 10);
    const prof = await prisma.user.upsert({
        where: { email: 'prof@eroz.com' },
        update: {},
        create: { email: 'prof@eroz.com', password: profPassword, firstName: 'Marie', lastName: 'Dupont', role: 'PROF' },
    });
    console.log('✅ Professor:', prof.email);
    await createSessionsAndStats(prof.id, 20, 10, prof.firstName);

    console.log('🎉 Done!');
}

main()
    .catch((e) => { console.error(e); process.exit(1); })
    .finally(async () => { await prisma.$disconnect(); });

File: backend/app/services/seed.py
from __future__ import annotations

import random
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from app.core.security import hash_password
from app.models import TrainingSession, User, UserStats

DIFFICULTIES = {
    "EASY": {"name": "EASY", "multiplier": 1.0},
    "MEDIUM": {"name": "MEDIUM", "multiplier": 1.5},
    "HARD": {"name": "HARD", "multiplier": 2.0},
}


def calculate_level(total_xp: int) -> int:
    return (total_xp // 1000) + 1


def calculate_streak(dates: list[datetime]) -> int:
    if not dates:
        return 0
    unique_days = sorted({d.date() for d in dates}, reverse=True)
    streak = 1
    for idx in range(1, len(unique_days)):
        if (unique_days[idx - 1] - unique_days[idx]).days == 1:
            streak += 1
        else:
            break
    return streak


def create_sessions_and_stats(
    db: Session,
    user: User,
    session_count: int,
    streak_days: int,
    user_name: str,
):
    sessions: list[TrainingSession] = []
    now = datetime.utcnow()
    total_xp = 0

    for i in range(session_count):
        if i < 7:
            days_ago = i
        elif i < session_count * 0.7:
            days_ago = random.randint(0, 6)
        else:
            days_ago = random.randint(7, 20)

        hours = random.randint(8, 19)
        minutes = random.randint(0, 59)

        completed_at = now - timedelta(days=days_ago)
        completed_at = completed_at.replace(hour=hours, minute=minutes, second=0, microsecond=0)

        diff_roll = random.random()
        if diff_roll < 0.35:
            difficulty = DIFFICULTIES["EASY"]
        elif diff_roll < 0.75:
            difficulty = DIFFICULTIES["MEDIUM"]
        else:
            difficulty = DIFFICULTIES["HARD"]

        total_images = random.randint(10, 20)
        correct_answers = random.randint(int(total_images * 0.6), int(total_images * 0.95))
        precision = round((correct_answers / total_images) * 100, 1)
        duration = random.randint(120, 420)

        base_score = int(precision * 10)
        xp_earned = round(base_score * difficulty["multiplier"])
        total_xp += xp_earned

        session = TrainingSession(
            userId=user.id,
            difficulty=difficulty["name"],
            precision=precision,
            duration=duration,
            totalImages=total_images,
            correctAnswers=correct_answers,
            baseScore=base_score,
            multiplier=difficulty["multiplier"],
            xpEarned=xp_earned,
            completedAt=completed_at,
        )
        db.add(session)
        sessions.append(session)

    db.commit()

    total_sessions = len(sessions)
    average_score = round(sum(s.precision for s in sessions) / total_sessions) if total_sessions else 0
    average_time = round(sum(s.duration for s in sessions) / total_sessions) if total_sessions else 0
    level = calculate_level(total_xp)

    stats = db.query(UserStats).filter(UserStats.userId == user.id).first()
    if stats:
        stats.totalXp = total_xp
        stats.level = level
        stats.totalSessions = total_sessions
        stats.averageScore = average_score
        stats.averageTime = average_time
        stats.currentStreak = streak_days
        stats.lastActivityAt = now
    else:
        stats = UserStats(
            userId=user.id,
            totalXp=total_xp,
            level=level,
            totalSessions=total_sessions,
            averageScore=average_score,
            averageTime=average_time,
            currentStreak=streak_days,
            lastActivityAt=now,
        )
        db.add(stats)

    db.commit()


def rebuild_stats_from_sessions(db: Session, user: User):
    sessions = (
        db.query(TrainingSession)
        .filter(TrainingSession.userId == user.id)
        .order_by(TrainingSession.completedAt.asc())
        .all()
    )
    if not sessions:
        return

    total_xp = sum(s.xpEarned for s in sessions)
    total_sessions = len(sessions)
    average_score = round(sum(s.precision for s in sessions) / total_sessions)
    average_time = round(sum(s.duration for s in sessions) / total_sessions)
    level = calculate_level(total_xp)
    last_activity = sessions[-1].completedAt
    streak = calculate_streak([s.completedAt for s in sessions])

    stats = db.query(UserStats).filter(UserStats.userId == user.id).first()
    if not stats:
        stats = UserStats(userId=user.id)
        db.add(stats)

    stats.totalXp = total_xp
    stats.level = level
    stats.totalSessions = total_sessions
    stats.averageScore = average_score
    stats.averageTime = average_time
    stats.currentStreak = streak
    stats.lastActivityAt = last_activity
    db.commit()


def seed_if_needed(db: Session):
    users_to_seed = [
        {
            "email": "admin@eroz.com",
            "password": "admin123",
            "firstName": "Admin",
            "lastName": "Eroz",
            "role": "ADMIN",
            "sessions": 30,
            "streak": 15,
        },
        {
            "email": "thomas.martin@edu.fr",
            "password": "student123",
            "firstName": "Thomas",
            "lastName": "Martin",
            "role": "STUDENT",
            "sessions": 12,
            "streak": 5,
        },
        {
            "email": "prof@eroz.com",
            "password": "prof123",
            "firstName": "Marie",
            "lastName": "Dupont",
            "role": "PROF",
            "sessions": 20,
            "streak": 10,
        },
    ]

    for entry in users_to_seed:
        user = db.query(User).filter(User.email == entry["email"]).first()
        if not user:
            user = User(
                email=entry["email"],
                password=hash_password(entry["password"]),
                firstName=entry["firstName"],
                lastName=entry["lastName"],
                role=entry["role"],
            )
            db.add(user)
            db.commit()
            db.refresh(user)

        has_sessions = (
            db.query(TrainingSession).filter(TrainingSession.userId == user.id).first() is not None
        )
        stats = db.query(UserStats).filter(UserStats.userId == user.id).first()

        if not has_sessions:
            create_sessions_and_stats(
                db,
                user,
                session_count=entry["sessions"],
                streak_days=entry["streak"],
                user_name=entry["firstName"],
            )
        elif not stats:
            rebuild_stats_from_sessions(db, user)
